subgraphs/_reference_/src/sdk/protocols/generic/pool.ts

```diff
@@ -0,0 +1,387 @@
+    this.tokens = tokens;
+    this.snapshoter = new PoolSnapshot(pool, protocol.event);
+  }
+
+  private save(): void {
+    this.pool.save();
+  }
+
+  private getInputToken(): Token {
+    return this.tokens.getOrCreateToken(
+      Address.fromBytes(this.pool.inputToken)
+    );
+  }
+
+  initialize(
+    name: string,
+    symbol: string,
+    inputToken: Token,
+    outputToken: Token | null,
+    isLiquidity: bool | null
+  ): void {
+    if (this.isInitialized) {
+      return;
+    }
+
+    const event = this.protocol.getCurrentEvent();
+    this.pool.protocol = this.protocol.getBytesID();
+    this.pool.name = name;
+    this.pool.symbol = symbol;
+    this.pool.inputToken = inputToken.id;
+    this.pool.outputToken = outputToken ? outputToken.id : null;
+    this.pool.isLiquidity = isLiquidity ? isLiquidity : null;
+    this.pool.createdTimestamp = event.block.timestamp;
+    this.pool.createdBlockNumber = event.block.number;
+
+    this.pool.inputTokenBalance = BIGINT_ZERO;
+    this.pool.totalValueLockedUSD = BIGDECIMAL_ZERO;
+    this.pool.cumulativeSupplySideRevenueUSD = BIGDECIMAL_ZERO;
+    this.pool.cumulativeProtocolSideRevenueUSD = BIGDECIMAL_ZERO;
+    this.pool.cumulativeTotalRevenueUSD = BIGDECIMAL_ZERO;
+
+    this.pool.lastSnapshotDayID = BIGINT_ZERO;
+    this.pool.lastSnapshotHourID = BIGINT_ZERO;
+    this.pool.lastUpdateTimestamp = BIGINT_ZERO;
+    this.save();
+
+    this.protocol.addPool();
+  }
+
+  /**
+   * Recalculates the total value locked for this pool based on its current input token balance.
+   * This function will also update the protocol's total value locked based on the change in this pool's.
+   */
+  refreshTotalValueLocked(): void {
+    const tvl = this.getInputTokenAmountPrice(this.pool.inputTokenBalance);
+    this.setTotalValueLocked(tvl);
+  }
+
+  /**
+   * Updates the total value locked for this pool to the given value.
+   * Will also update the protocol's total value locked based on the change in this pool's.
+   */
+  setTotalValueLocked(newTVL: BigDecimal): void {
+    const delta = newTVL.minus(this.pool.totalValueLockedUSD);
+    this.addTotalValueLocked(delta);
+    this.save();
+  }
+
+  /**
+   * Adds the given delta to the total value locked for this pool.
+   * Will also update the protocol's total value locked based on the change in this pool's.
+   *
+   * @param delta The change in total value locked for this pool.
+   */
+  addTotalValueLocked(delta: BigDecimal): void {
+    this.pool.totalValueLockedUSD = this.pool.totalValueLockedUSD.plus(delta);
+    this.protocol.addTotalValueLocked(delta);
+    this.save();
+  }
+
+  /**
+   * Utility function to convert some amount of input token to USD.
+   *
+   * @param amount the amount of inputToken to convert to USD
+   * @returns The converted amount.
+   */
+  getInputTokenAmountPrice(amount: BigInt): BigDecimal {
+    const token = this.getInputToken();
+    const price = this.protocol.getTokenPricer().getTokenPrice(token);
```

Comment:
Note to self: For some reason I added this same method to the TokenPricer interface. For convenience, to not have to do the calculation every time. And at the same time, I added it to the `Pool` too 😅 

I think we could remove the `getAmountValueUSD()` method from the TokenPricer interface. Since the SDK can calculate it from `getTokenPrice` alone, and this way we save the client from implementing it every time.